### PR TITLE
Updated proposal services

### PR DESCRIPTION
**Pull request by chnixi**
Adapted create_proposal, modify_proposal, add_sus, subtract_sus in account_logic.py as shown in the UML diagram of the architecture file. Changed the name "kwargs" to "clusters_sus" to make it sound more descriptive. Added boolean "all_clusters" variable, which is however not used yet.

### CODE REVIEW

File: bank/account_logic.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class ProposalServices:
 
     def __init__(self, account_name: str) -> None:
         """Administrate proposal data for the given user account
-
+
```
Comment (djperrefort) [inline bank/account_logic.py:32]: Drop trailing whitespace (Use a linter)

File: bank/account_logic.py [hunk 2]
```diff
@@ -165,25 +164,32 @@ def modify_proposal(
...
         pid = pid or self._get_active_pid()
         self._verify_proposal_id(pid)
-        self._verify_cluster_values(**kwargs)
-
+        self._verify_cluster_values(**clusters_sus)
+
+        # Validate start and end times
+        if not (start or end):
+            raise ValueError(f'modify_date requires either a new start: {start} or new end: {end} date')
+        if (start and end) and start > end:
```
Comment (djperrefort) [inline bank/account_logic.py:190]: Use `>=`.

**Review by djperrefort — CHANGES_REQUESTED** [review]
Only a few minor comments. Otherwise looks good.

File: bank/account_logic.py (hunk 1 above)
Comment (chnixi) [reply bank/account_logic.py]: My bad. Forgot how massively sensitive Python is to indentation, making it unfriendly for a pure plain text editor.

File: bank/account_logic.py (hunk 2 above)
Comment (chnixi) [reply bank/account_logic.py]: Fixed.

**Review by djperrefort — APPROVED** [review]

File: bank/account_logic.py [hunk 3]
```diff
@@ -215,58 +225,61 @@ def modify_proposal(
...
+            LOG.info(f"Modified proposal {proposal.id} for account {self._account_name}. Overwrote {clusters_sus}")
+
+    def add_sus(self, pid: Optional[int] = None, **clusters_sus: int) -> None:
         """Add service units to an account proposal
 
         Args:
             pid: Modify a specific proposal by its inv_id (Defaults to currently active proposal)
-            **kwargs: Service units to add for each cluster
-
+            all_clusters: Specifies if all clusters are affected
```
Comment (Comeani) [inline bank/account_logic.py:239]: I think all_clusters will be encapsulated in `**cluster_sus`, we'll want to include some conditional logic here for making sure `cluster_sus` is updated from `{all_clusters:SUS}` to `{SMP:SUS, MPI:SUS, GPU:SUS ... }`. The behavior should probably be to only apply the `all_clusters` value if it's provided. ex. `--all_clusters 100 --SMP 150` -> `--all_clusters 100` only.

Comment (Comeani) [reply bank/account_logic.py:239]: Perhaps we could do this in `_verify_cluster_values`, having it update `**cluster_sus` with it's return value.

File: bank/account_logic.py [hunk 4]
```diff
@@ -215,58 +225,61 @@ def modify_proposal(
...
+            **clusters_sus: Service units to add for each cluster
         Raises:
             MissingProposalError: If the proposal ID does not match the account
         """
 
         pid = pid or self._get_active_pid()
         self._verify_proposal_id(pid)
-        self._verify_cluster_values(**kwargs)
+        self._verify_cluster_values(**clusters_sus)
```
Comment (Comeani) [inline bank/account_logic.py:272]: See comment above about all_clusters

File: bank/account_logic.py [hunk 5]
```diff
@@ -349,7 +361,7 @@ def create(
             self,
             sus: int,
             start: Optional[date] = date.today(),
-            end: Optional[date] = None,
+            end: Optional[date] = date.today() + relativedelta(years=1),
```
Comment (Comeani) [inline bank/account_logic.py:364]: Change this to `None` and do something below like `end = end or (start + relativedelta(years=1))`, similar adjustment to other functions that take optional start/end

File: bank/cli.py [hunk 6]
```diff
@@ -243,12 +243,14 @@ def define_interface(cls, parent_parser) -> None:
             help='The ID associated with a specific proposal on the account'
         )
 
+
```
Comment (Comeani) [inline bank/cli.py:246]: extra line here

File: bank/settings.py [hunk 7]
```diff
@@ -91,6 +91,8 @@
 # Settings for running the test suite.
 test_account = 'sam'
 test_cluster = 'smp'
+non_existent_account = 'dummy'
```
Comment (Comeani) [inline bank/settings.py:94]: testing with an instance of https://github.com/giovtorres/slurm-docker-cluster until our environments have some test accounts in the DB

File: tests/account_logic/test_ProposalServices.py [hunk 8]
```diff
@@ -22,8 +22,17 @@
     .where(Proposal.is_active)
 
 
+class InitExceptions(EmptyAccountSetup, TestCase):
+    """Tests to ensure proposals report that provided account does not exist"""
+
+    def setUp(self) -> None:
```
Comment (Comeani) [inline tests/account_logic/test_ProposalServices.py:28]: change to `test_error_on_non_existent_account`

File: tests/account_logic/test_ProposalServices.py [hunk 9]
```diff
@@ -268,22 +238,25 @@ def setUp(self) -> None:
...
         su_kwargs = {settings.test_cluster: 1}
-        self.account.create_proposal(start=YESTERDAY, duration=2, **su_kwargs)
-        self.account.create_proposal(start=TOMORROW, duration=2, **su_kwargs)
+
+        self.account.create(start=YESTERDAY, end=YESTERDAY+timedelta(days=2), **{settings.test_cluster: 100})
+        self.account.create(start=TOMORROW, end=TOMORROW+timedelta(days=2), **{settings.test_cluster: 100})
 
         with self.assertRaises(ProposalExistsError):
-            self.account.modify_proposal(end=DAY_AFTER_TOMORROW)
+            self.account.modify_date(end=DAY_AFTER_TOMORROW)
```
Comment (Comeani) [inline tests/account_logic/test_ProposalServices.py:262]: Double check that I didn't remove EOF newlines elsewhere...

**Review by Comeani — CHANGES_REQUESTED** [review]
Notes to self after taking a pass through recent changes

File: bank/account_logic.py (hunk 3 above)
Comment (Comeani) [reply bank/account_logic.py]: Realizing my explanation above isn't quite how it should work, we want the SU amount specified for `--all_clusters` available across **any** of the clusters, not awarded to each of them. It's also possible someone might want to award to a specific cluster **and** give floating units (that are not an investment) in a single create/add/subtract call. Resolving this for now, and we'll consider how it needs to be interacted with in locking with `update_account_status`

**Pull request merged**